### PR TITLE
[Whisper] Fix word-level timestamps with bs>1 or num_beams>1

### LINKED ISSUE #27362: Mismatched dimensions error when using beam_size != 1 and return_timestamps="word" for AutomaticSpeechRecognitionPipeline with Whisper
### System Info

- `transformers` version: 4.33.3
- Platform: Linux-5.15.0-88-generic-x86_64-with-glibc2.31
- Python version: 3.9.18
- Huggingface_hub version: 0.17.3
- Safetensors version: 0.3.3
- Accelerate version: 0.23.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@sanchit-gandhi

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I have the following code:
```
model_name_or_path = "openai/whisper-small"
LANGUAGE = "French"
TASK = "transcribe"
CHUNK_LENGTH_S = 30  # 30 seconds chunks
SAMPLE_RATE = 16_000

model = WhisperForConditionalGeneration.from_pretrained(
    pretrained_model_name_or_path=model_name_or_path,
    load_in_8bit=True,
    device_map="cuda:0",
)

feature_extractor = WhisperFeatureExtractor.from_pretrained(
    model_name_or_path,
)
tokenizer = WhisperTokenizer.from_pretrained(
    model_name_or_path,
    language=LANGUAGE,
    task=TASK,
)
processor = WhisperProcessor.from_pretrained(
    model_name_or_path,
    language=LANGUAGE,
    task=TASK,
)

pipeline_generator = AutomaticSpeechRecognitionPipeline(
    model=model,
    tokenizer=tokenizer,
    feature_extractor=feature_extractor,
    chunk_length_s=CHUNK_LENGTH_S,
    stride_length_s=5
)
pipeline_generator.model.config.forced_decoder_ids = (
    pipeline_generator.tokenizer.get_decoder_prompt_ids(
        language=LANGUAGE, task=TASK
    )
)

forced_decoder_ids = processor.get_decoder_prompt_ids(
    language=LANGUAGE, task=TASK
)

model.config.use_cache = True
model.eval()

transcript = pipeline_generator(
    inputs=inputs,
    return_timestamps="word",
    generate_kwargs={
        "num_beams": 5,
    },
)
```

After executing the code, I receive the following error:
```
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 356, in __call__
    return super().__call__(inputs, **kwargs)
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/pipelines/base.py", line 1132, in __call__
    return next(
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/pipelines/pt_utils.py", line 124, in __next__
    item = next(self.iterator)
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/pipelines/pt_utils.py", line 266, in __next__
    processed = self.infer(next(self.iterator), **self.params)
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/pipelines/base.py", line 1046, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 551, in _forward
    tokens = self.model.generate(
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/models/whisper/modeling_whisper.py", line 1768, in generate
    outputs["token_timestamps"] = self._extract_token_timestamps(outputs, generation_config.alignment_heads)
  File "/mnt/disk2/nestor/batvoice/dev/bv-etl-transcribe/env/lib/python3.9/site-packages/transformers/models/whisper/modeling_whisper.py", line 1834, in _extract_token_timestamps
    timestamps[batch_idx, 1:] = torch.tensor(jump_times)
RuntimeError: The expanded size of the tensor (105) must match the existing size (135) at non-singleton dimension 0.  Target sizes: [105].  Tensor sizes: [135]
```

### Expected behavior

When I change the parameter num_beams to 1, it works perfectly.

Is there any incompatibility between these two parameters? I've looked in the documentation and I haven't been able to find so.

The problem seems to be in this function (transformers/models/whisper/modeling_whisper.py):
```
    def _extract_token_timestamps(self, generate_outputs, alignment_heads, time_precision=0.02):
        """
        Calculates token-level timestamps using the encoder-decoder cross-attentions and dynamic time-warping (DTW) to
        map each output token to a position in the input audio.

        Returns:
            tensor containing the timestamps in seconds for each predicted token
        """
        # Create a list with `decoder_layers` elements, each a tensor of shape
        # (batch size, attention_heads, output length, input length).
        cross_attentions = []
        for i in range(self.config.decoder_layers):
            cross_attentions.append(torch.cat([x[i] for x in generate_outputs.cross_attentions], dim=2))

        # Select specific cross-attention layers and heads. This is a tensor
        # of shape (batch size, num selected, output length, input length).
        weights = torch.stack([cross_attentions[l][:, h] for l, h in alignment_heads])
        weights = weights.permute([1, 0, 2, 3])

        # Normalize and smoothen the weights.
        std, mean = torch.std_mean(weights, dim=-2, keepdim=True, unbiased=False)
        weights = (weights - mean) / std
        weights = _median_filter(weights, self.config.median_filter_width)

        # Average the different cross-attention heads.
        matrix = weights.mean(dim=1)

        timestamps = torch.zeros_like(generate_outputs.sequences, dtype=torch.float32)

        # Perform dynamic time warping on each element of the batch.
        for batch_idx in range(timestamps.shape[0]):
            text_indices, time_indices = _dynamic_time_warping(-matrix[batch_idx].double().cpu().numpy())
            jumps = np.pad(np.diff(text_indices), (1, 0), constant_values=1).astype(bool)
            jump_times = time_indices[jumps] * time_precision
            timestamps[batch_idx, 1:] = torch.tensor(jump_times)

        return timestamps
```

Thank you in advance.

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

Supersedes #26699

This PR fixes two issues related to Whisper:
1. Wrong DTW matrix computation when computing word-level timestamps with beam search (issues #27362 and #28007)
2. Bug when computing world-levels timestamps with bs>1 using the pipeline (issue #27446 and PR #26699)

The first issue happens because the DTW matrix is derived from the cross attentions. The latter is of size `beam_search*num_return_sequences*batch_size`,  but it should be of size `num_return_sequences*batch_size` , so we need to keep track of the beam indices.

The second issue happens because when batching with the pipeline, `stride` is passed as a list of tuple (one per sample) instead of a single tuple.
When there are multiple strides passed to `_extract_token_timestamps`, we can't compute the DTW matrix in parallel.
It is treated in two cases:

1. If same stride for each sample, compute DTW weights in parallel
2. If not the same stride (i.e end of audio file) compute them sequentially

The loss of parallelism is not so dramatic, since in all cases the DTW algorithm is performed sequentially.

Fixes #27362, #28007, #27446 

cc @sanchit-gandhi, @amyeroberts

### CODE REVIEW

File: tests/models/whisper/test_modeling_whisper.py
Comment (ylacombe) [inline tests/models/whisper/test_modeling_whisper.py:1878]: Not super fan of not being able to compare it to the ground-truth, like done [here](https://github.com/huggingface/transformers/blob/892dd91edde5ad8e7ba47ab35778b179b527a59f/tests/models/whisper/test_modeling_whisper.py#L1843-L1848). I've tried to compute it with the original whisper but it doesn't seem straightforward.

Any ideas on how to easily compute ground-truth here @sanchit-gandhi

File: src/transformers/models/whisper/modeling_whisper.py
Comment (ylacombe) [inline src/transformers/models/whisper/modeling_whisper.py:2581]: Could you check that this makes sense @gante ?

Basically, this reconciles `generate_outputs.cross_attentions` with `generate_outputs.sequences` thanks to `generate_outputs.beam_indices`.

File: src/transformers/models/whisper/modeling_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:2611]: Good for me smoothing the cross-attention weights matrix sequentially (rather than batched) since we have to do it this way for DTW anyway

File: src/transformers/models/whisper/modeling_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:2567]: (nit)
```suggestion
            # If beam index is still -1, it means that the associated token id is EOS
```

File: tests/models/whisper/test_modeling_whisper.py
Comment (sanchit-gandhi) [reply tests/models/whisper/test_modeling_whisper.py:1878]: In the interest of time, what you can do is batched generation with bs=2, and then simply confirm that the outputs from the first batch match what we got with the un-batched integration test [mentioned above](https://github.com/huggingface/transformers/blob/892dd91edde5ad8e7ba47ab35778b179b527a59f/tests/models/whisper/test_modeling_whisper.py#L1843-L1848)

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py
Comment (sanchit-gandhi) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:718]: The trailing comma after `EXPECTED_OUTPUT,` is causing the linter to split this over 3 lines instead of 1

```suggestion
        self.assertDictEqual(output, EXPECTED_OUTPUT)
```

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py
Comment (sanchit-gandhi) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:725]:
```suggestion
        self.assertDictEqual(output, EXPECTED_OUTPUT)
```

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for fixing this @ylacombe!

File: src/transformers/models/whisper/modeling_whisper.py
Comment (amyeroberts) [inline src/transformers/models/whisper/modeling_whisper.py:2562]: I think?
```suggestion
            # Since the beam search strategy chooses the most probable sequences at the end of the search
```

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:721]: Just to make sure I've understood here. In this case, `sample` will effectively be loaded twice in the dataloader to create a batch_size of 2. 

Is there a reason for not using two different samples i.e. `sample = data[:2]["audio"]` ?

File: src/transformers/models/whisper/modeling_whisper.py
Comment (amyeroberts) [inline src/transformers/models/whisper/modeling_whisper.py:2567]: I'm not super au fait with beam generation but the comment and code here aren't clear to me:
* What do we mean 'still -1'. If we've generated EOS won't it remain -1? 
* 'associated token id' - associated with what? 
* What does setting to 0 here mean?

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Most comments are for my own understanding of what happening. 

Have you run the slow tests to confirm generation with whisper is backwards compatible with previous behaviour?

File: src/transformers/pipelines/automatic_speech_recognition.py
Comment (ylacombe) [reply src/transformers/pipelines/automatic_speech_recognition.py:565]: TBH, I'd have to check that what you proposed works with other models, as those lines are specific to Whisper

File: tests/models/whisper/test_modeling_whisper.py
Comment (ylacombe) [reply tests/models/whisper/test_modeling_whisper.py:1864]: I've simply reproduced the codes from the other `test_tiny_...` tests!

File: tests/models/whisper/test_modeling_whisper.py
Comment (ylacombe) [reply tests/models/whisper/test_modeling_whisper.py:1878]: The issue here is to test if the beam-search code snippet is working, not the batch generation ! So the problem can't be resolved with what you mention right ?

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py
Comment (ylacombe) [reply tests/pipelines/test_pipelines_automatic_speech_recognition.py:718]: If I understood correctly, if `chunk_length_s<len(audio)`, then the audio is split into a certain number of chunks, which then computed in a batch if  `batch_size>1`.

So in any case, this test if the batch feature works correctly!
 

cc @sanchit-gandhi, is that correct ?

**Review by ylacombe — COMMENTED** [review]
Hey @amyeroberts and @sanchit-gandhi, many thanks for reviewing this!

As these are really used features, I want to make sure that we align here before modifying the tests and modeling code. 
So I simply answered some of your questions (alongside with nit modifs) here, let me know what you think of this

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py
Comment (amyeroberts) [reply tests/pipelines/test_pipelines_automatic_speech_recognition.py:718]: Ah, OK, makes sense. Could you add a comment explaining that here? Otherwise it's not obvious what the test is doing

File: tests/models/whisper/test_modeling_whisper.py
Comment (amyeroberts) [reply tests/models/whisper/test_modeling_whisper.py]: OK!

File: src/transformers/pipelines/automatic_speech_recognition.py
Comment (amyeroberts) [reply src/transformers/pipelines/automatic_speech_recognition.py:565]: OK, in this case, let's leave as-is for the moment

**Review by amyeroberts — APPROVED** [review]
Thanks for working on this and answering all my questions! 

Happy to merge once all slow whisper tests (incl. modeling and generation) have been run and pass

File: src/transformers/models/whisper/modeling_whisper.py
Comment (ylacombe) [inline src/transformers/models/whisper/modeling_whisper.py:2579]: Thanks @amy for the review! All of the tests related to Whisper or the pipeline are green!

I'd still think @gante's opinion of this piece of code is needed before merging though!

**Comment by ylacombe** [comment]
Merging now, once the checks are green, as it is a blocking feature, all tests are passing and @gante is off for a few days!

**Comment by guy1992l** [comment]
Hi @ylacombe 
Thanks for the fix! I do still face an issue I think is relevant here so I figured I'll post it.

When I'm trying to use the whisper model with batch size > 1 (for my example: 2) and word level timestamps, I get this error in the DTW method:

```
...
File "/transformers/pipelines/base.py", line 926, in device_placement
    yield
File "/transformers/pipelines/base.py", line 1046, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
File "/transformers/pipelines/automatic_speech_recognition.py", line 572, in _forward
    tokens = self.model.generate(
File "/transformers/models/whisper/modeling_whisper.py", line 2273, in generate
    outputs["token_timestamps"] = self._extract_token_timestamps(
File "/transformers/models/whisper/modeling_whisper.py", line 2624, in _extract_token_timestamps
    text_indices, time_indices = _dynamic_time_warping(-matrix.double().cpu().numpy())
File "/transformers/models/whisper/modeling_whisper.py", line 258, in _dynamic_time_warping
    output_length, input_length = matrix.shape
ValueError: too many values to unpack (expected 2)
```

The `matrix` object's shape is `(6, 72, 1500)` and indeed it is expecting a shape of 2 so it fails. Do you have any idea what happened or how can I overcome it?

Thanks!

**Pull request merged**
